### DISCUSSION TITLE
Implementing schema-based multi-tenancy with NestJS-Query modules

### DISCUSSION THREAD

**Discussion by mariusorani** [TriPSs/nestjs-query#303] [Q&A]
## Background
I'm developing a multi-tenant application using NestJS with NestJS-Query modules. Specifically, I need to implement schema-based multi-tenancy, where each tenant has its own database schema. I'm having difficulty integrating this approach with NestJS-Query.

## What I've Researched
I've looked into:
1. Creating dynamic database connections based on tenant
2. Using custom connection factories
3. Modifying NestJS-Query services to use tenant-specific connections

However, I'm unsure how to properly implement these ideas with NestJS-Query, which abstracts much of the database interaction.

## My Questions
1. Has anyone successfully implemented schema-based multi-tenancy with NestJS-Query modules?
2. Are there any examples or open-source projects demonstrating this specific approach?
3. How can I modify NestJS-Query services to connect to the correct tenant schema for each request?
4. What's the best way to handle database migrations across multiple tenant schemas when using NestJS-Query?
5. Are there any performance considerations or best practices for schema-based multi-tenancy with NestJS-Query?

## Additional Context
- I'm using TypeORM as my ORM
- Each tenant should have its own isolated schema
- I need to dynamically switch between schemas based on the incoming request
- I want to ensure that NestJS-Query's features (like automatic CRUD operations) still work in this multi-schema environment

Any guidance, code snippets, or links to relevant resources would be immensely helpful. I've searched extensively but haven't found clear examples of NestJS-Query being used in a schema-based multi-tenant application.

**Comment by GP4cK** [comment]
I can only help give a partial answer as I have worked with multi-tenancy but all the tenants had the same schema:

First, you can create a class that implements `TypeOrmOptionsFactory`:

```ts
@Injectable()
export class TypeOrmConfigFactory implements TypeOrmOptionsFactory {
  private readonly tenantID: string;
  // Here you can inject the request to extract the tenant id (from the JWT for example)
  constructor(@Inject(REQUEST) req: Request) {
    tenantID = ...
  }

  createTypeOrmOptions() {
    // Build your ORM config here with the tenant ID.
    //! Do not give a name here, or you will have to do so in all the calls to TypeOrm.forFeature.
    return {
      database: tenantID,
      // ...
    }
  }
}
```


Then in your AppModule (or wherever you're configuring TypeOrm), you can do this:

```ts
@Module({
  imports: [
    TypeOrmModule.forRootAsync({
      useClass: TypeOrmConfigService,
    }),
    // ...
  ],
})
export class AppModule implements NestModule { /* ... */ }
```

Now, nestjs-query will automatically use the right connection.
However, since your schemas are tenant specific, you may not  be able to use the default resolvers of nestjs-query but you may take a look at the decorators that are used to generate the classes to achieve what you want.
I hope that helps at least a little.

**Reply by mariusorani** [reply]
Thank's a lot. At least this is a good starting point. The default resolvers may work but i think i have to separate the subscription and default resolvers, since subscription resolvers can't be request scoped
